fast_llm/engine/multi_stage/fast_llm_model.py

```diff
@@ -83,28 +85,83 @@ def distributed(self):
 
     def save_checkpoint(
         self,
-        checkpoint_config: CheckpointConfig,
+        checkpoint_config: CheckpointSaveConfig,
         metadata: dict | None = None,
     ):
-        if metadata is None:
-            metadata = {}
-        if checkpoint_config.checkpoint_type == CheckpointType.distributed:
-            self._save_distributed_checkpoint(checkpoint_config, metadata)
-        elif checkpoint_config.checkpoint_type == CheckpointType.state_dict:
-            self._save_state_dict_checkpoint(checkpoint_config, metadata)
-        elif checkpoint_config.checkpoint_type == CheckpointType.huggingface:
-            assert checkpoint_config.exported_model_type is not None
-            self._export_checkpoint(checkpoint_config, metadata)
+        # TODO: Handle barriers, ok file, mkdir, etc. here
+
+        num_shards = len(self._state_shard_names) if checkpoint_config.optimizer_state else 1
+        metadata = {
+            "checkpoint_type": CheckpointFormat.distributed.value,
+            "checkpoint_version": str(CHECKPOINT_VERSION),
+            "fast_llm_config": self._fast_llm_config.to_serialized(),
+            "state_shard_names": list(self._state_shard_names[:num_shards]),
+            "metadata": {} if metadata is None else metadata,
+        }
+
+        # TODO: Simplify branching.
+        if checkpoint_config.format == CheckpointFormat.external:
+            # TODO: Support optimizer?
```

Comment:
can you confirm that Fast-LLM will set the `checkpoint_config.optimizer_state` to `False` somewhere (or reject the config at launch) before we end up here? Would be frustrating to crash thousands of steps into training because of this.

Reply:
Right now there is no safety check, but that would be easy to add. Looking at it, right now the bigger problem is that the optimizer isn't saved by default so the bigger risk is not saving it when we want to...